Platform/OpenNtfsDxe/Data.c

```diff
@@ -0,0 +1,566 @@
+
+  ASSERT(FileSystem != NULL);
+  ASSERT(FileSystem->DiskIo != NULL);
+  ASSERT(FileSystem->BlockIo != NULL);
+
+  Media = FileSystem->BlockIo->Media;
+
+  Status = FileSystem->DiskIo->ReadDisk (
+    FileSystem->DiskIo,
+    Media->MediaId,
+    Offset,
+    Size,
+    Buffer
+    );
+  if (EFI_ERROR(Status)) {
+    DEBUG((DEBUG_INFO, "NTFS: Could not read disk at address %08x\n", Offset));
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+ReadMftRecord (
+  IN  EFI_NTFS_FILE *File,
+  OUT UINT8         *Buffer,
+  IN  UINT64        RecordNumber
+  )
+{
+  EFI_STATUS     Status;
+
+  Status = ReadAttr (
+    &File->MftFile.Attr,
+    Buffer,
+    RecordNumber * mFileRecordSize,
+    mFileRecordSize
+    );
+  if (EFI_ERROR(Status)) {
+    DEBUG((DEBUG_INFO, "NTFS: Could not read MFT Record 0x%Lx\n", RecordNumber));
+    return Status;
+  }
+
+  return Fixup (Buffer, mFileRecordSize, "FILE");
+}
+
+EFI_STATUS
+EFIAPI
+ReadAttr (
+  IN  NTFS_ATTR *Attr,
+  OUT UINT8     *Dest,
+  IN  UINT64    Offset,
+  IN  UINTN     Length
+  )
+{
+  EFI_STATUS       Status;
+  UINT8            *Current;
+  UINT32           Type;
+  UINT8            *AttrStart;
+  UINT64           Vcn;
+  ATTR_LIST_RECORD *Record;
+  ATTR_HEADER_RES  *Res;
+
+  Current = Attr->Current;
+  Attr->Next = Attr->Current;
+
+  if ((Attr->Flags & NTFS_AF_ALST) != 0) {
+    if (Attr->Last < (Attr->Next + sizeof (ATTR_LIST_RECORD))) {
+      DEBUG((DEBUG_INFO, "NTFS: $ATTRIBUTE_LIST does not contain even a single record.\n"));
+      return EFI_VOLUME_CORRUPTED;
+    }
+
+    Record = (ATTR_LIST_RECORD *) Attr->Next;
+    Type = Record->Type;
+
+    Vcn = DivU64x64Remainder (Offset, mClusterSize, NULL);
+    if (COMPRESSION_BLOCK >= mClusterSize) {
+      Vcn &= ~0xFULL;
+    }
+
+    Record = (ATTR_LIST_RECORD *) ((UINT8 *)Record + Record->RecordLength);
+    while (((UINT8 *)Record + sizeof (ATTR_LIST_RECORD)) <= Attr->Last) {
+      if (Record->Type != Type) {
+        break;
+      }
+
+      if (Record->StartingVCN > Vcn) {
+        break;
+      }
+
+      if (Record->RecordLength == 0) {
+        return EFI_VOLUME_CORRUPTED;
+      }
+
+      Attr->Next = (UINT8 *) Record;
+      Record = (ATTR_LIST_RECORD *) ((UINT8 *) Record + Record->RecordLength);
+    }
+  } else {
+    Res = (ATTR_HEADER_RES *) Attr->Next;
+    Type = Res->Type;
+  }
+
+  AttrStart = FindAttr (Attr, Type);
+  if (AttrStart != NULL) {
+    Status = ReadData (Attr, AttrStart, Dest, Offset, Length);
+  } else {
+    DEBUG((DEBUG_INFO, "NTFS: Attribute not found\n"));
+    Status = EFI_VOLUME_CORRUPTED;
+  }
+
+  Attr->Current = Current;
+
+  return Status;
+}
+
+EFI_STATUS
+EFIAPI
+ReadData (
+  IN  NTFS_ATTR *Attr,
+  IN  UINT8     *AttrStart,
+  OUT UINT8     *Dest,
+  IN  UINT64    Offset,
+  IN  UINTN     Length
+  )
+{
+  EFI_STATUS         Status;
+  RUNLIST            *Runlist;
+  ATTR_HEADER_NONRES *NonRes;
+  ATTR_HEADER_RES    *Res;
+  // UINT64             Sector0;
```

Comment:
Remove these declarations if no longer needed.

Reply:
I am not sure, what exactly commented out pieces of code do. They have never been used in any of my tests, but there may exist some special cases, where this code may become necessary.

Reply:
I am not sure the origin of the commented out code... but we are basically saying that since we do not know what `NTFS_AF_GPOS` does, and have never hit it in any tests, we are removing all code which references it?

Is there any way to get clearer about what the attribute is, and when it might be set?